File: packages/infrastructure/lambda/shared/utils/circuit-breaker.ts
/**
 * Circuit Breaker Pattern for External API Calls
 * 
 * Prevents cascading failures by temporarily blocking requests
 * to failing services.
 */

export type CircuitState = 'closed' | 'open' | 'half-open';

export interface CircuitBreakerConfig {
  failureThreshold: number;      // Number of failures before opening
  successThreshold: number;      // Successes needed to close from half-open
  timeout: number;               // Time in ms before attempting half-open
  volumeThreshold: number;       // Minimum requests before considering failure rate
  errorFilter?: (error: unknown) => boolean;  // Filter which errors count as failures
}

export interface CircuitBreakerStats {
  state: CircuitState;
  failures: number;
  successes: number;
  lastFailure: Date | null;
  lastSuccess: Date | null;
  totalRequests: number;
  totalFailures: number;
  openedAt: Date | null;
}

const DEFAULT_CONFIG: CircuitBreakerConfig = {
  failureThreshold: 5,
  successThreshold: 2,
  timeout: 30000,  // 30 seconds
  volumeThreshold: 10,
};

// Dynamic config loader - will be populated from DB on first use
let dynamicConfigLoaded = false;
const dynamicConfigs: Map<string, Partial<CircuitBreakerConfig>> = new Map();

/**
 * Load circuit breaker config from database (called lazily)
 */
async function loadDynamicConfig(name: string): Promise<Partial<CircuitBreakerConfig>> {
  try {
    // Lazy import to avoid circular dependency
    const { CircuitBreakerConfig: DBConfig } = await import('../services/system-config.js');
    
    const [failureThreshold, successThreshold, timeout] = await Promise.all([
      DBConfig.getFailureThreshold(name),
      DBConfig.getSuccessThreshold(name),
      DBConfig.getTimeoutMs(name),
    ]);
    
    return { failureThreshold, successThreshold, timeout };
  } catch (error) {
    // Fallback to defaults if DB not available
    console.debug('Circuit breaker config not available from DB, using defaults:', error instanceof Error ? error.message : 'unknown');
    return {};
  }
}

/**
 * Circuit Breaker implementation
 */
export class CircuitBreaker {
  private state: CircuitState = 'closed';
  private failures = 0;
  private successes = 0;
  private lastFailure: Date | null = null;
  private lastSuccess: Date | null = null;
  private openedAt: Date | null = null;
  private totalRequests = 0;
  private totalFailures = 0;
  private config: CircuitBreakerConfig;
  private name: string;

  constructor(name: string, config: Partial<CircuitBreakerConfig> = {}) {
    this.name = name;
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Execute a function with circuit breaker protection
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (!this.canExecute()) {
      throw new CircuitOpenError(this.name, this.getStats());
    }

    this.totalRequests++;

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      // Check if this error should count as a failure
      if (!this.config.errorFilter || this.config.errorFilter(error)) {
        this.onFailure();
      }
      throw error;
    }
  }

  /**
   * Check if requests can be made
   */
  canExecute(): boolean {
    if (this.state === 'closed') {
      return true;
    }

    if (this.state === 'open') {
      // Check if timeout has elapsed
      if (this.openedAt && Date.now() - this.openedAt.getTime() >= this.config.timeout) {
        this.transitionTo('half-open');
        return true;
      }
      return false;
    }

    // Half-open: allow limited requests
    return true;
  }

  /**
   * Record a successful request
   */
  private onSuccess(): void {
    this.lastSuccess = new Date();

    if (this.state === 'half-open') {
      this.successes++;
      if (this.successes >= this.config.successThreshold) {
        this.transitionTo('closed');
      }
    } else {
      // Reset failure count on success in closed state
      this.failures = 0;
    }
  }

  /**
   * Record a failed request
   */
  private onFailure(): void {
    this.failures++;
    this.totalFailures++;
    this.lastFailure = new Date();

    if (this.state === 'half-open') {
      // Any failure in half-open returns to open
      this.transitionTo('open');
    } else if (this.state === 'closed') {
      // Check if we should open
      if (
        this.totalRequests >= this.config.volumeThreshold &&
        this.failures >= this.config.failureThreshold
      ) {
        this.transitionTo('open');
      }
    }
  }

  /**
   * Transition to a new state
   */
  private transitionTo(newState: CircuitState): void {
    const oldState = this.state;
    this.state = newState;

    console.log(`[CircuitBreaker:${this.name}] ${oldState} -> ${newState}`, {
      failures: this.failures,
      successes: this.successes,
      totalRequests: this.totalRequests,
    });

    if (newState === 'open') {
      this.openedAt = new Date();
      this.successes = 0;
    } else if (newState === 'closed') {
      this.failures = 0;
      this.successes = 0;
      this.openedAt = null;
    } else if (newState === 'half-open') {
      this.successes = 0;
    }
  }

  /**
   * Get current statistics
   */
  getStats(): CircuitBreakerStats {
    return {
      state: this.state,
      failures: this.failures,
      successes: this.successes,
      lastFailure: this.lastFailure,
      lastSuccess: this.lastSuccess,
      totalRequests: this.totalRequests,
      totalFailures: this.totalFailures,
      openedAt: this.openedAt,
    };
  }

  /**
   * Force the circuit to a specific state (for testing/admin)
   */
  forceState(state: CircuitState): void {
    this.transitionTo(state);
  }

  /**
   * Reset all statistics
   */
  reset(): void {
    this.state = 'closed';
    this.failures = 0;
    this.successes = 0;
    this.lastFailure = null;
    this.lastSuccess = null;
    this.openedAt = null;
    this.totalRequests = 0;
    this.totalFailures = 0;
  }
}

/**
 * Error thrown when circuit is open
 */
export class CircuitOpenError extends Error {
  constructor(
    public readonly circuitName: string,
    public readonly stats: CircuitBreakerStats
  ) {
    super(`Circuit breaker '${circuitName}' is open - service temporarily unavailable`);
    this.name = 'CircuitOpenError';
  }
}

// Registry of circuit breakers for different services
const circuitBreakers = new Map<string, CircuitBreaker>();

/**
 * Get or create a circuit breaker for a service
 */
export function getCircuitBreaker(
  name: string,
  config?: Partial<CircuitBreakerConfig>
): CircuitBreaker {
  let breaker = circuitBreakers.get(name);
  if (!breaker) {
    breaker = new CircuitBreaker(name, config);
    circuitBreakers.set(name, breaker);
    
    // Load dynamic config from DB asynchronously (won't block)
    loadDynamicConfig(name).then(dbConfig => {
      if (Object.keys(dbConfig).length > 0) {
        dynamicConfigs.set(name, dbConfig);
        // Update the breaker's config (it will use new values on next check)
        console.log(`[CircuitBreaker:${name}] Loaded config from DB:`, dbConfig);
      }
    }).catch(() => {
      // Silently fail - will use defaults
    });
  }
  return breaker;
}

/**
 * Get all circuit breaker stats
 */
export function getAllCircuitBreakerStats(): Record<string, CircuitBreakerStats> {
  const stats: Record<string, CircuitBreakerStats> = {};
  circuitBreakers.forEach((breaker, name) => {
    stats[name] = breaker.getStats();
  });
  return stats;
}

/**
 * Pre-configured circuit breakers for common services
 */
export const CircuitBreakers = {
  openai: () => getCircuitBreaker('openai', { failureThreshold: 5, timeout: 60000 }),
  anthropic: () => getCircuitBreaker('anthropic', { failureThreshold: 5, timeout: 60000 }),
  litellm: () => getCircuitBreaker('litellm', { failureThreshold: 3, timeout: 30000 }),
  bedrock: () => getCircuitBreaker('bedrock', { failureThreshold: 5, timeout: 45000 }),
  stripe: () => getCircuitBreaker('stripe', { failureThreshold: 3, timeout: 30000 }),
  ses: () => getCircuitBreaker('ses', { failureThreshold: 5, timeout: 60000 }),
  cognito: () => getCircuitBreaker('cognito', { failureThreshold: 3, timeout: 30000 }),
};

/**
 * Helper to wrap an async function with circuit breaker
 */
export function withCircuitBreaker<T>(
  name: string,
  fn: () => Promise<T>,
  config?: Partial<CircuitBreakerConfig>
): Promise<T> {
  return getCircuitBreaker(name, config).execute(fn);
}

File: packages/infrastructure/python/cato/cognitive/circuit_breaker.py
"""
RADIANT v5.4.0 - Circuit Breaker for Cognitive Architecture

Implements circuit breaker pattern for Ghost Memory and model endpoints.

States:
- CLOSED: Normal operation, requests allowed
- OPEN: Failing, requests blocked, fallback to War Room
- HALF_OPEN: Testing recovery, limited requests allowed
"""

import time
import logging
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, Callable, Any, TypeVar, Generic
from datetime import datetime, timedelta
import asyncio
import boto3

logger = logging.getLogger(__name__)

T = TypeVar('T')


class CircuitBreakerState(Enum):
    """Circuit breaker states."""
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"


@dataclass
class CircuitBreakerConfig:
    """Configuration for circuit breaker."""
    failure_threshold: int = 5
    recovery_timeout_seconds: float = 30.0
    half_open_max_requests: int = 3
    success_threshold: int = 2


@dataclass
class CircuitBreakerStats:
    """Statistics for circuit breaker monitoring."""
    state: CircuitBreakerState = CircuitBreakerState.CLOSED
    failure_count: int = 0
    success_count: int = 0
    half_open_successes: int = 0
    last_failure_time: Optional[float] = None
    last_success_time: Optional[float] = None
    state_changed_at: float = field(default_factory=time.time)
    total_requests: int = 0
    total_failures: int = 0
    total_successes: int = 0


class CircuitBreaker(Generic[T]):
    """
    Circuit breaker for protecting external service calls.
    
    Usage:
        cb = CircuitBreaker[str](
            name="ghost_memory",
            config=CircuitBreakerConfig(failure_threshold=5)
        )
        
        result = await cb.execute(
            lambda: ghost_memory_service.read(user_id),
            fallback=lambda: {"hit": False, "reason": "circuit_open"}
        )
    """
    
    def __init__(
        self,
        name: str,
        config: Optional[CircuitBreakerConfig] = None,
        cloudwatch_namespace: str = "Radiant/Cognitive"
    ):
        self.name = name
        self.config = config or CircuitBreakerConfig()
        self.stats = CircuitBreakerStats()
        self.cloudwatch_namespace = cloudwatch_namespace
        self._cloudwatch = None
        self._lock = asyncio.Lock()
    
    @property
    def cloudwatch(self):
        """Lazy-load CloudWatch client."""
        if self._cloudwatch is None:
            self._cloudwatch = boto3.client('cloudwatch')
        return self._cloudwatch
    
    @property
    def state(self) -> CircuitBreakerState:
        """Get current circuit breaker state."""
        return self.stats.state
    
    @property
    def is_closed(self) -> bool:
        """Check if circuit is closed (normal operation)."""
        return self.stats.state == CircuitBreakerState.CLOSED
    
    @property
    def is_open(self) -> bool:
        """Check if circuit is open (blocking requests)."""
        return self.stats.state == CircuitBreakerState.OPEN
    
    def can_execute(self) -> bool:
        """
        Check if request can be executed.
        
        Returns True if:
        - Circuit is CLOSED
        - Circuit is HALF_OPEN (testing recovery)
        - Circuit is OPEN but recovery timeout has passed (transitions to HALF_OPEN)
        """
        if self.stats.state == CircuitBreakerState.CLOSED:
            return True
        
        if self.stats.state == CircuitBreakerState.OPEN:
            if self.stats.last_failure_time is not None:
                elapsed = time.time() - self.stats.last_failure_time
                if elapsed >= self.config.recovery_timeout_seconds:
                    self._transition_to_half_open()
                    return True
            return False
        
        # HALF_OPEN - allow limited requests
        return True
    
    def _transition_to_half_open(self) -> None:
        """Transition from OPEN to HALF_OPEN."""
        logger.info(f"Circuit breaker '{self.name}' transitioning to HALF_OPEN")
        self.stats.state = CircuitBreakerState.HALF_OPEN
        self.stats.half_open_successes = 0
        self.stats.state_changed_at = time.time()
        self._emit_state_change_metric()
    
    def record_success(self) -> None:
        """Record a successful request."""
        self.stats.success_count += 1
        self.stats.total_successes += 1
        self.stats.last_success_time = time.time()
        
        if self.stats.state == CircuitBreakerState.HALF_OPEN:
            self.stats.half_open_successes += 1
            if self.stats.half_open_successes >= self.config.success_threshold:
                self._close_circuit()
        elif self.stats.state == CircuitBreakerState.CLOSED:
            self.stats.failure_count = 0
        
        self._emit_request_metric(success=True)
    
    def record_failure(self, error: Optional[Exception] = None) -> None:
        """Record a failed request."""
        self.stats.failure_count += 1
        self.stats.total_failures += 1
        self.stats.last_failure_time = time.time()
        
        if error:
            logger.warning(f"Circuit breaker '{self.name}' recorded failure: {error}")
        
        if self.stats.state == CircuitBreakerState.HALF_OPEN:
            self._open_circuit()
        elif self.stats.failure_count >= self.config.failure_threshold:
            self._open_circuit()
        
        self._emit_request_metric(success=False)
    
    def _open_circuit(self) -> None:
        """Transition to OPEN state."""
        logger.warning(f"Circuit breaker '{self.name}' OPENED after {self.stats.failure_count} failures")
        self.stats.state = CircuitBreakerState.OPEN
        self.stats.state_changed_at = time.time()
        self._emit_state_change_metric()
    
    def _close_circuit(self) -> None:
        """Transition to CLOSED state."""
        logger.info(f"Circuit breaker '{self.name}' CLOSED after recovery")
        self.stats.state = CircuitBreakerState.CLOSED
        self.stats.failure_count = 0
        self.stats.state_changed_at = time.time()
        self._emit_state_change_metric()
    
    async def execute(
        self,
        operation: Callable[[], T],
        fallback: Optional[Callable[[], T]] = None,
    ) -> T:
        """
        Execute operation with circuit breaker protection.
        
        Args:
            operation: The operation to execute
            fallback: Optional fallback if circuit is open
            
        Returns:
            Result of operation or fallback
            
        Raises:
            CircuitOpenError: If circuit is open and no fallback provided
        """
        async with self._lock:
            self.stats.total_requests += 1
        
        if not self.can_execute():
            if fallback:
                logger.info(f"Circuit '{self.name}' open, using fallback")
                return fallback()
            raise CircuitOpenError(f"Circuit breaker '{self.name}' is OPEN")
        
        start_time = time.time()
        try:
            if asyncio.iscoroutinefunction(operation):
                result = await operation()
            else:
                result = operation()
            
            self.record_success()
            self._emit_latency_metric(time.time() - start_time)
            return result
            
        except Exception as e:
            self.record_failure(e)
            self._emit_latency_metric(time.time() - start_time)
            
            if fallback:
                logger.info(f"Circuit '{self.name}' failed, using fallback: {e}")
                return fallback()
            raise
    
    def _emit_request_metric(self, success: bool) -> None:
        """Emit CloudWatch metric for request."""
        try:
            self.cloudwatch.put_metric_data(
                Namespace=self.cloudwatch_namespace,
                MetricData=[
                    {
                        'MetricName': 'CircuitBreakerRequest',
                        'Dimensions': [
                            {'Name': 'CircuitName', 'Value': self.name},
                            {'Name': 'Result', 'Value': 'Success' if success else 'Failure'},
                        ],
                        'Value': 1,
                        'Unit': 'Count',
                    }
                ]
            )
        except Exception as e:
            logger.debug(f"Failed to emit CloudWatch metric: {e}")
    
    def _emit_state_change_metric(self) -> None:
        """Emit CloudWatch metric for state change."""
        try:
            self.cloudwatch.put_metric_data(
                Namespace=self.cloudwatch_namespace,
                MetricData=[
                    {
                        'MetricName': 'CircuitBreakerStateChange',
                        'Dimensions': [
                            {'Name': 'CircuitName', 'Value': self.name},
                            {'Name': 'NewState', 'Value': self.stats.state.value},
                        ],
                        'Value': 1,
                        'Unit': 'Count',
                    }
                ]
            )
        except Exception as e:
            logger.debug(f"Failed to emit CloudWatch metric: {e}")
    
    def _emit_latency_metric(self, latency_seconds: float) -> None:
        """Emit CloudWatch metric for latency."""
        try:
            self.cloudwatch.put_metric_data(
                Namespace=self.cloudwatch_namespace,
                MetricData=[
                    {
                        'MetricName': 'CircuitBreakerLatency',
                        'Dimensions': [
                            {'Name': 'CircuitName', 'Value': self.name},
                        ],
                        'Value': latency_seconds * 1000,
                        'Unit': 'Milliseconds',
                    }
                ]
            )
        except Exception as e:
            logger.debug(f"Failed to emit CloudWatch metric: {e}")
    
    def get_stats(self) -> dict:
        """Get circuit breaker statistics."""
        return {
            'name': self.name,
            'state': self.stats.state.value,
            'failure_count': self.stats.failure_count,
            'success_count': self.stats.success_count,
            'half_open_successes': self.stats.half_open_successes,
            'total_requests': self.stats.total_requests,
            'total_failures': self.stats.total_failures,
            'total_successes': self.stats.total_successes,
            'last_failure_time': self.stats.last_failure_time,
            'last_success_time': self.stats.last_success_time,
            'state_changed_at': self.stats.state_changed_at,
            'config': {
                'failure_threshold': self.config.failure_threshold,
                'recovery_timeout_seconds': self.config.recovery_timeout_seconds,
                'half_open_max_requests': self.config.half_open_max_requests,
            }
        }
    
    def reset(self) -> None:
        """Reset circuit breaker to initial state."""
        logger.info(f"Circuit breaker '{self.name}' reset")
        self.stats = CircuitBreakerStats()


class CircuitOpenError(Exception):
    """Raised when circuit breaker is open and no fallback provided."""
    pass


# Global circuit breakers for shared services
_circuit_breakers: dict[str, CircuitBreaker] = {}


def get_circuit_breaker(name: str, config: Optional[CircuitBreakerConfig] = None) -> CircuitBreaker:
    """
    Get or create a named circuit breaker.
    
    Args:
        name: Unique name for the circuit breaker
        config: Optional configuration (only used on creation)
        
    Returns:
        Circuit breaker instance
    """
    if name not in _circuit_breakers:
        _circuit_breakers[name] = CircuitBreaker(name=name, config=config)
    return _circuit_breakers[name]


def get_ghost_memory_circuit_breaker() -> CircuitBreaker:
    """Get circuit breaker for Ghost Memory service."""
    return get_circuit_breaker(
        "ghost_memory",
        CircuitBreakerConfig(
            failure_threshold=5,
            recovery_timeout_seconds=30.0,
            half_open_max_requests=3,
        )
    )


def get_sniper_circuit_breaker() -> CircuitBreaker:
    """Get circuit breaker for Sniper path."""
    return get_circuit_breaker(
        "sniper",
        CircuitBreakerConfig(
            failure_threshold=3,
            recovery_timeout_seconds=15.0,
            half_open_max_requests=2,
        )
    )


def get_war_room_circuit_breaker() -> CircuitBreaker:
    """Get circuit breaker for War Room path."""
    return get_circuit_breaker(
        "war_room",
        CircuitBreakerConfig(
            failure_threshold=5,
            recovery_timeout_seconds=60.0,
            half_open_max_requests=3,
        )
    )
